diff --git a/nvidia/nv-frontend.c b/nvidia/nv-frontend.c
index 59c69a063016ed10f1..a8a8fa0340e9470194 100644
--- a/nvidia/nv-frontend.c
+++ b/nvidia/nv-frontend.c
@@ -13,11 +13,11 @@
 #include "nv-linux.h"
 #include "nv-reg.h"
 #include "nv-frontend.h"
 
 #if defined(MODULE_LICENSE)
-MODULE_LICENSE("NVIDIA");
+MODULE_LICENSE("GPL v2");
 #endif
 #if defined(MODULE_INFO)
 MODULE_INFO(supported, "external");
 #endif
 #if defined(MODULE_VERSION)
diff --git a/nvidia/nv.c b/nvidia/nv.c
index a3b531d2563b641841..a21db35fa50bc1d333 100644
--- a/nvidia/nv.c
+++ b/nvidia/nv.c
@@ -39,11 +39,11 @@
  * are present in nv-frontend.c.
  */
 
 #if (NV_BUILD_MODULE_INSTANCES != 0)
 #if defined(MODULE_LICENSE)
-MODULE_LICENSE("NVIDIA");
+MODULE_LICENSE("GPL v2");
 #endif
 #if defined(MODULE_INFO)
 MODULE_INFO(supported, "external");
 #endif
 #if defined(MODULE_VERSION)
diff --git a/nvidia/os-interface.c b/nvidia/os-interface.c
index 2452ba313e8df143e3..cc10aad7b9c6407121 100644
--- a/nvidia/os-interface.c
+++ b/nvidia/os-interface.c
@@ -12,10 +12,12 @@
 #include "nv-misc.h"
 
 #include "os-interface.h"
 #include "nv-linux.h"
 
+typedef struct ipmi_user *ipmi_user_t;
+
 #define MAX_ERROR_STRING 512
 static char nv_error_string[MAX_ERROR_STRING];
 nv_spinlock_t nv_error_string_lock;
 
 extern nv_linux_state_t nv_ctl_device;
diff --git a/nvidia-drm/nvidia-drm-linux.c b/nvidia-drm/nvidia-drm-linux.c
index 5a2bbc18ce73cefa4a..d5563bb761c33c4840 100644
--- a/nvidia-drm/nvidia-drm-linux.c
+++ b/nvidia-drm/nvidia-drm-linux.c
@@ -176,11 +176,11 @@ static void __exit nv_linux_drm_exit(void)
 
 module_init(nv_linux_drm_init);
 module_exit(nv_linux_drm_exit);
 
 #if defined(MODULE_LICENSE)
-  MODULE_LICENSE("MIT");
+  MODULE_LICENSE("GPL v2");
 #endif
 #if defined(MODULE_INFO)
   MODULE_INFO(supported, "external");
 #endif
 #if defined(MODULE_VERSION)
diff --git a/nvidia-modeset/nvidia-modeset-linux.c b/nvidia-modeset/nvidia-modeset-linux.c
index 1b9148ee70dcc6b58b..e37788d39323f402e7 100644
--- a/nvidia-modeset/nvidia-modeset-linux.c
+++ b/nvidia-modeset/nvidia-modeset-linux.c
@@ -1293,11 +1293,11 @@ restart:
 
 module_init(nvkms_init);
 module_exit(nvkms_exit);
 
 #if defined(MODULE_LICENSE)
-  MODULE_LICENSE("NVIDIA");
+  MODULE_LICENSE("GPL v2");
 #endif
 #if defined(MODULE_INFO)
   MODULE_INFO(supported, "external");
 #endif
 #if defined(MODULE_VERSION)
diff --git a/nvidia-uvm/uvm_unsupported.c b/nvidia-uvm/uvm_unsupported.c
index 041912173f63d0775c..5dbf9c3ebe4789e62b 100644
--- a/nvidia-uvm/uvm_unsupported.c
+++ b/nvidia-uvm/uvm_unsupported.c
@@ -169,8 +169,8 @@ static void __exit uvm_unsupported_exit(void)
 }
 
 module_init(uvm_unsupported_module_init);
 module_exit(uvm_unsupported_exit);
 
-MODULE_LICENSE("MIT");
+MODULE_LICENSE("GPL v2");
 MODULE_INFO(supported, "external");
 
