diff --git a/cmake/FindNcurses.cmake b/cmake/FindNcurses.cmake
index cc2f17126..774dd02d0 100644
--- a/cmake/FindNcurses.cmake
+++ b/cmake/FindNcurses.cmake
@@ -47,6 +47,12 @@ else()
   endif()
 endif()
 
+find_package(PkgConfig QUIET)
+If(PKG_CONFIG_FOUND)
+  pkg_search_module(NCURSES ncurses)
+  set(NCURSES_LIBRARY ${LIBS} ${NCURSES_LIBRARIES})
+endif()
+
 if(NCURSES_INCLUDE_PATH AND NCURSES_LIBRARY)
   set(NCURSES_FOUND TRUE)
 endif()
