diff --git a/jellyshot b/jellyshot
index cee169b..0cc78bb 100755
--- a/jellyshot
+++ b/jellyshot
@@ -4,9 +4,9 @@
 set -E -o pipefail
 
 # Traps for cleanup
-trap '{ [[ -f "$file" ]] && rm -v "$file"; }' EXIT
-trap '{ [[ -f "$file" ]] && rm -v "$file"; exit 1; }' ERR
-trap '{ [[ -f "$file" ]] && rm -v "$file"; trap - INT; kill -INT $$; }' INT
+trap '{ tput setaf 9; [[ -f "$file" ]] && rm -v "$file"; tput sgr0; }' EXIT
+trap '{ tput setaf 9; [[ -f "$file" ]] && rm -v "$file"; tput sgr0; exit 1; }' ERR
+trap '{ tput setaf 9; [[ -f "$file" ]] && rm -v "$file"; tput sgr0; trap - INT; kill -INT $$; }' INT
 
 function rand_element () {
   local -a arr
@@ -19,7 +19,7 @@ config="${XDG_CONFIG_HOME:-${HOME}/.config}/jellyshot/config"
 if [[ -f "$config" ]]; then
   . "$config"
 else
-  printf "%s: %s\n" "No configuration available at" "$config"
+  printf "%s: $(tput setaf 9)%s$(tput sgr0)\n" "No configuration available at" "$config"
   exit 1
 fi
 
@@ -37,7 +37,7 @@ file="/tmp/$name"
 if type import &>/dev/null; then
   import "$file"
 else
-  printf "%s\n" '"import" not found in $PATH'
+  printf "$(tput setaf 9)%s$(tput sgr0)\n" '"import" not found in $PATH'
   exit 1
 fi
 
@@ -60,9 +60,9 @@ else
 fi
 
 if type xclip &>/dev/null; then
-  echo "$url" && printf "$url" | xclip -i -selection clipboard
+  printf "$(tput setaf 10)%s$(tput sgr0)\n" "$url" && printf "$url" | xclip -i -selection clipboard
 elif type xsel &>/dev/null; then
-  echo "$url" && printf "$url" | xsel -ib
+  printf "$(tput setaf 10)%s$(tput sgr0)\n" "$url" && printf "$url" | xsel -ib
 else
   printf "%s\n" "$url"
 fi
